src/com/amazon/ionschema/AuthorityFilesystem.kt

```diff
@@ -28,15 +29,29 @@ import java.io.Reader
  *
  * @property[basePath] The base path in the filesystem in which to resolve schema identifiers.
  */
-class AuthorityFilesystem(private val basePath: String) : Authority {
+class AuthorityFilesystem(basePath: String) : Authority {
+    private val basePath: String
+
     init {
-        if (!File(basePath).exists()) {
-            throw IllegalArgumentException("Path '$basePath' does not exist")
+        val file = File(basePath)
+        if (!file.exists()) {
+            throw FileNotFoundException("Path '$basePath' does not exist")
+        }
+
+        this.basePath = if (file.canonicalPath.endsWith(File.separator)) {
+            file.canonicalPath
+        } else {
+            file.canonicalPath + File.separator
         }
     }
 
     override fun iteratorFor(iss: IonSchemaSystem, id: String): CloseableIterator<IonValue> {
         val file = File(basePath, id)
+
+        if (!file.canonicalPath.startsWith(basePath)) {
+            throw FileNotFoundException("$id (Permission denied)")
```

Comment:
Would [`AccessDeniedException`](https://docs.oracle.com/javase/7/docs/api/java/nio/file/AccessDeniedException.html) be a better fit for this?

There are some circumstances where `FileNotFoundException` doesn't strictly mean that the file wasn't found (e.g. [if you try to open a read-only file for writing](https://docs.oracle.com/javase/8/docs/api/index.html?java/io/FileNotFoundException.html)), but that would feel pretty surprising to me as a user.

Reply:
Agreed, AccessDeniedException is a better fit; opting for [kotlin.io.AccessDeniedException](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/-access-denied-exception/) instead of the Java equivalent.  Thank you :)